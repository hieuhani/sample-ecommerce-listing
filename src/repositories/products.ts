import { api } from '../services/api'

export interface ProductSearchParams {
  saleCategories?: number;
  channel?: string;
  terminal?: string;
  saleStatuses?: string;
  _sort?: string;
  _order?: 'asc' | 'desc';
  _page?: number;
  _limit?: number;
  publishStatus?: boolean;
  q?: string;
}

export interface ProductPrice {
  supplierSalePrice: number;
  sellPrice: number;
}

export interface ProductPromotionPrice {
  channel: string;
  terminal: string;
  finalPrice: number;
  promotionPrice?: number;
  bestPrice: number;
  flashSalePrice?: number;
}

export interface ProductImage {
  url: string;
  priority: number;
}

export interface Product {
  sku: string;
  displayName: string;
  price: ProductPrice;
  promotionPrices: Array<ProductPromotionPrice>;
  images: Array<ProductImage>;
}

export interface SearchResponseExtra {
  totalItems: number;
  page: number;
  pageSize: number;
}

export interface ProductSearchResponse {
  result: {
    products: Array<Product>;
  };
  extra: SearchResponseExtra;
}

const defaultSearchParams: ProductSearchParams = {
  saleCategories: 613,
  channel: 'pv_online',
  terminal: 'phongvu',
  saleStatuses: 'hang_ban,hang_dat_truoc,hang_sap_het,hang_moi,hang_trung_bay,hang_thanh_ly',
  _sort: 'saleStatuses||hang_ban|hang_dat_truoc|hang_sap_het|hang_moi||hang_trung_bay|hang_thanh_ly||ngung_kinh_doanh',
  _order: 'asc',
  _page: 1,
  _limit: 20,
  publishStatus: true,
}

export function search(params: ProductSearchParams): Promise<ProductSearchResponse> {
  return api.get('/search', Object.assign({}, defaultSearchParams, params)).then(({ ok, data, problem }) => {
    if (!ok) {
      return Promise.reject(problem)
    }
    return Promise.resolve(data as ProductSearchResponse)
  })
}
