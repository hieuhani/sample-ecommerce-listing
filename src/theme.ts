import { DefaultTheme } from 'styled-components'

export const theme: DefaultTheme = {
  colors: {
    reddishOrange: '#f5471e',
    coolGrey: '#8f9598',
    silver: '#c6cace',
    tomato: '#ea341f',
    darkGrey: '#262829',
    paleGrey: '#eef1f3',
    white: '#FFF',
    black30: 'rgba(0, 0, 0, .3)',
    pumpkinOrange: '#f58400',
    deepSkyBlue: '#1e75f6',
  }
}
