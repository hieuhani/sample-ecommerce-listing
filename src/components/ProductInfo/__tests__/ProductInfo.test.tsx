import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { ProductInfo } from '..'
import { theme } from '../../../theme'
import { Product } from '../../../repositories/products'

const product: Product = {
  displayName: 'Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)',
  sku: '1807564',
  images: [{
    url: 'https://phongvu.vn/media/catalog_v2/media/48/18/1570775921.343882_Acer_AS_XC-885_1.jpg',
    priority: 1,
  }],
  status: {
    publish: true,
    sale: 'ngung_kinh_doanh' as const,
  },
  promotionPrices: [{
    channel: 'pv_online',
    terminal: 'phongvu',
    finalPrice: 11890000,
    bestPrice: 11890000,
    promotionPrice: 5000000,
  }],
  price: {
    supplierSalePrice: 11890000,
    sellPrice: 11890000,
  },
  saleCategories: [],
  attributeGroups: [],
  attributes: [],
}

describe('ProductInfo', () => {
  test('ProductInfo renders correctly', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <ProductInfo product={product} />
      </ThemeProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('ProductInfo renders correctly when sale status is not valid', () => {
    const invalidSaleStatusProduct = Object.assign({}, product, {
      status: {
        publish: true,
        sale: 'invalid_status',
      },
    })
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <ProductInfo product={invalidSaleStatusProduct} />
      </ThemeProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('ProductInfo renders correctly when there are no promotion', () => {
    const noPromotionProduct = Object.assign({}, product, {
      promotionPrices: [],
    })
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <ProductInfo product={noPromotionProduct} />
      </ThemeProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
