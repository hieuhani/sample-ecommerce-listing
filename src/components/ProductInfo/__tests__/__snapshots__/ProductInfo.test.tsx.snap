// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductInfo ProductInfo renders correctly 1`] = `
.c8 {
  top: -.3em;
  left: .1em;
}

.c10 {
  color: #8f9598;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 6pt;
}

.c11 span {
  color: #FFF;
  background: rgb(232,32,32);
  background: linear-gradient(90deg,rgba(232,32,32,1) 0%,#f5471e 100%);
  padding: 2pt 2pt 2pt 0;
  border-top-right-radius: 2pt;
  border-bottom-right-radius: 2pt;
  font-size: 0.8rem;
  margin-left: -6pt;
  z-index: 1001;
}

.c11:before {
  z-index: 1000;
  content: ' ';
  width: 12.3pt;
  height: 12.3pt;
  background: rgb(232,32,32);
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  border-bottom-left-radius: 2pt;
}

.c0 {
  padding: 12pt;
  background-color: #FFF;
  border-top: 1px solid #eef1f3;
  margin-bottom: 8pt;
}

.c1 {
  font-size: 17pt;
  color: #262829;
}

.c4 {
  color: #1e75f6;
}

.c3 {
  color: #8f9598;
}

.c5 {
  background-color: rgb(238,241,243);
  color: #8f9598;
  font-size: 12pt;
  padding: 3pt 8pt;
  border-radius: 8pt;
}

.c2 {
  margin-bottom: 8pt;
}

.c6 {
  margin-top: 8pt;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  font-size: 17pt;
  font-weight: 700;
  color: #ea341f;
  margin-bottom: 0;
}

.c9 {
  width: 8pt;
}

<div
  className="c0"
>
  <h3
    className="c1"
  >
    Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)
  </h3>
  <div
    className="c2"
  >
    <span
      className="c3"
    >
      Mã SP: 
    </span>
    <span
      className="c4"
    >
      1807564
    </span>
  </div>
  <div>
    <span
      className="c5"
    >
      Ngừng kinh doanh
    </span>
  </div>
  <div
    className="c6"
  >
    <h3
      className="c7"
    >
      11.890.000
      <sup
        className="c8"
      >
        đ
      </sup>
    </h3>
    <span
      className="c9"
    />
    <span
      className="c10"
    >
      11.890.000
    </span>
    <span
      className="c9"
    />
    <div
      className="c11"
    >
      <span>
        -
        58
        %
      </span>
    </div>
  </div>
</div>
`;

exports[`ProductInfo ProductInfo renders correctly when sale status is not valid 1`] = `
.c8 {
  top: -.3em;
  left: .1em;
}

.c10 {
  color: #8f9598;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 6pt;
}

.c11 span {
  color: #FFF;
  background: rgb(232,32,32);
  background: linear-gradient(90deg,rgba(232,32,32,1) 0%,#f5471e 100%);
  padding: 2pt 2pt 2pt 0;
  border-top-right-radius: 2pt;
  border-bottom-right-radius: 2pt;
  font-size: 0.8rem;
  margin-left: -6pt;
  z-index: 1001;
}

.c11:before {
  z-index: 1000;
  content: ' ';
  width: 12.3pt;
  height: 12.3pt;
  background: rgb(232,32,32);
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  border-bottom-left-radius: 2pt;
}

.c0 {
  padding: 12pt;
  background-color: #FFF;
  border-top: 1px solid #eef1f3;
  margin-bottom: 8pt;
}

.c1 {
  font-size: 17pt;
  color: #262829;
}

.c4 {
  color: #1e75f6;
}

.c3 {
  color: #8f9598;
}

.c5 {
  background-color: rgb(238,241,243);
  color: #8f9598;
  font-size: 12pt;
  padding: 3pt 8pt;
  border-radius: 8pt;
}

.c2 {
  margin-bottom: 8pt;
}

.c6 {
  margin-top: 8pt;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  font-size: 17pt;
  font-weight: 700;
  color: #ea341f;
  margin-bottom: 0;
}

.c9 {
  width: 8pt;
}

<div
  className="c0"
>
  <h3
    className="c1"
  >
    Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)
  </h3>
  <div
    className="c2"
  >
    <span
      className="c3"
    >
      Mã SP: 
    </span>
    <span
      className="c4"
    >
      1807564
    </span>
  </div>
  <div>
    <span
      className="c5"
    >
      invalid_status
    </span>
  </div>
  <div
    className="c6"
  >
    <h3
      className="c7"
    >
      11.890.000
      <sup
        className="c8"
      >
        đ
      </sup>
    </h3>
    <span
      className="c9"
    />
    <span
      className="c10"
    >
      11.890.000
    </span>
    <span
      className="c9"
    />
    <div
      className="c11"
    >
      <span>
        -
        58
        %
      </span>
    </div>
  </div>
</div>
`;

exports[`ProductInfo ProductInfo renders correctly when there are no promotion 1`] = `
.c8 {
  top: -.3em;
  left: .1em;
}

.c0 {
  padding: 12pt;
  background-color: #FFF;
  border-top: 1px solid #eef1f3;
  margin-bottom: 8pt;
}

.c1 {
  font-size: 17pt;
  color: #262829;
}

.c4 {
  color: #1e75f6;
}

.c3 {
  color: #8f9598;
}

.c5 {
  background-color: rgb(238,241,243);
  color: #8f9598;
  font-size: 12pt;
  padding: 3pt 8pt;
  border-radius: 8pt;
}

.c2 {
  margin-bottom: 8pt;
}

.c6 {
  margin-top: 8pt;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  font-size: 17pt;
  font-weight: 700;
  color: #ea341f;
  margin-bottom: 0;
}

<div
  className="c0"
>
  <h3
    className="c1"
  >
    Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)
  </h3>
  <div
    className="c2"
  >
    <span
      className="c3"
    >
      Mã SP: 
    </span>
    <span
      className="c4"
    >
      1807564
    </span>
  </div>
  <div>
    <span
      className="c5"
    >
      Ngừng kinh doanh
    </span>
  </div>
  <div
    className="c6"
  >
    <h3
      className="c7"
    >
      11.890.000
      <sup
        className="c8"
      >
        đ
      </sup>
    </h3>
  </div>
</div>
`;
